index.js

```diff
@@ -49,12 +49,95 @@ function createWindow () {
     win.show()
   })
 
+  win.webContents.on('before-reload', async (event) => {
+    // Prevent the default reload behavior
+    event.preventDefault()
+    
+    try {
+      // Tell renderer to do cleanup
+      win.webContents.send('cleanup-before-reload')
+      
+      // Wait for cleanup then reload
+      setTimeout(() => {
+        // This will trigger a page reload, but won't trigger 'before-reload' again
+        win.reload()
+      }, 500)
+    } catch(e) {
+      console.error('Reload preparation failed:', e)
+    }
+  })
+
+  const initialMenuState = {
+    isConnected: false,
+    view: 'editor'
+  }
+
   registerIPCHandlers(win, ipcMain, app, dialog)
-  registerMenu(win)
+  registerMenu(win, initialMenuState)
 
   app.on('activate', () => {
     if (BrowserWindow.getAllWindows().length === 0) createWindow()
   })
 }
 
-app.on('ready', createWindow)
+function shortcutAction(key) {
+  win.webContents.send('shortcut-cmd', key);
+}
+
+// Shortcuts
+function registerShortcuts() {
+  globalShortcut.register('CommandOrControl+R', () => {
+    console.log('Running Program')
+    shortcutAction('r')
+  })
+  globalShortcut.register('CommandOrControl+Alt+R', () => {
+    console.log('Running Code Selection')
+    shortcutAction('_r')
+  })
+  globalShortcut.register('CommandOrControl+H', () => {
+    console.log('Stopping Program (Halt)')
+    shortcutAction('h')
+  })
+  globalShortcut.register('CommandOrControl+S', () => {
+    console.log('Saving File')
+    shortcutAction('s')
+  })
+  
+  globalShortcut.register('CommandOrControl+Shift+R', () => {
+    console.log('Resetting Board')
+    shortcutAction('R')
+  })
+  globalShortcut.register('CommandOrControl+Shift+C', () => {
+    console.log('Connect to Board')
+    shortcutAction('C')
+  })
+  globalShortcut.register('CommandOrControl+Shift+D', () => {
+    console.log('Disconnect from Board')
+    shortcutAction('D')
+  }),
+  globalShortcut.register('CommandOrControl+K', () => {
+    console.log('Clear Terminal')
+    shortcutAction('K')
+  }),
+  // Future: Toggle REPL Panel
+  // globalShortcut.register('CommandOrControl+T', () => {
+  //   console.log('Toggle Terminal')
+  //   shortcutAction('T')
+  // }),
+  globalShortcut.register('Escape', () => {
+    shortcutAction('ESC')
+  })
+}
+
+app.on('ready', () => {
+  createWindow()
+  registerShortcuts()
+
+  win.on('focus', () => {
+    registerShortcuts()
+  })
+  win.on('blur', () => {
+    globalShortcut.unregisterAll()
+  })
```

Comment:
I haven't looked into the details of the shortcut handling. But it looks like a hack registering system wide global shortcuts when focussing a window and unregistering them when blurring. Why can't we use the default accelerator handling?

Reply:
this will require further investigation, I'd leave it for another future PR